### PR TITLE
Connects Falkonry/tracker#815

### PR DESCRIPTION
**Pull request by dilpeshjain**

### CODE REVIEW

File: test/FalkonryTests.cs [hunk 1]
```diff
@@ -173,15 +174,15 @@ public void createDatastreamWideFormatSingleEntity()
             var inputList = new List<Input>();
             var currents = new Input();
             currents.Name = "current";
-            currents.ValueType = new ValueType();
+            currents.ValueType = new helper.models.ValueType();
```
Comment (aniket-amrutkar) [inline test/FalkonryTests.cs]: why we need this change ?

File: test/FalkonryTests.cs [hunk 2]
```diff
@@ -760,14 +761,122 @@ public void TestAssessmentCRUD()
         }
     }
 
-    // [TestClass()]
+     [TestClass()]
     public class TestAddHistoricalData
     {
-        Falkonry _falkonry = new Falkonry("https://localhost:8080", "auth-token");
+        Falkonry _falkonry1 = new Falkonry("https://localhost:8080", "auth-token");
```
Comment (aniket-amrutkar) [inline test/FalkonryTests.cs]: rename it to _falkonry

File: test/FalkonryTests.cs (hunk 1 above)
Comment (dilpeshjain) [reply test/FalkonryTests.cs]: Was facing some issue in importing value type file, late got resolved. Will remove it.

File: test/FalkonryTests.cs [hunk 3]
```diff
@@ -901,12 +1008,15 @@ public void AddDataCsv()
...
                 Assert.AreEqual(ds.DataSource.Type, datastream.DataSource.Type);
-                var data = "time, Unit, current, vibration, state\n 2016-05-05T12:00:00Z, Unit1, 12.4, 3.4, On";
+                var data = "time,Unit,current,vibration,state\n 2016-05-05T12:00:00.000Z,Unit1,12.4,3.4,On";
                 var options = new SortedDictionary<string, string>();
                 options.Add("timeIdentifier", "time");
                 options.Add("timeFormat", "iso_8601");
+                options.Add("timeZone", time.Zone);
                 options.Add("streaming", "false");
                 options.Add("hasMoreData", "false");
+                options.Add("entityIdentifier", "Unit");
```
Comment (aniket-amrutkar) [inline test/FalkonryTests.cs]: this test case is for single entity datastream. 
Why are we sending entityIdentifier ?

File: test/FalkonryTests.cs [hunk 4]
```diff
@@ -760,14 +761,122 @@ public void TestAssessmentCRUD()
...
+                //_falkonry.DeleteDatastream(datastream.Id);
+            }
+            catch (System.Exception exception)
+            {
+                Assert.AreEqual(exception.Message, null, false);
+            }
+
+        }
+
+        // Add wide input data (json format) to single thing Datastream
```
Comment (aniket-amrutkar) [inline test/FalkonryTests.cs]: rename thing to entity everywhere

File: test/FalkonryTests.cs [hunk 5]
```diff
@@ -917,28 +1027,77 @@ public void AddDataCsv()
...
+            }
+            catch (System.Exception exception)
+            {
+                Assert.AreEqual(exception.Message, "Missing time identifier.", true);
+            }
+
+        }
+
+        [TestMethod()]
+        // Add historical input data(csv format) to Datastream with time identifier missing
```
Comment (aniket-amrutkar) [inline test/FalkonryTests.cs]: It should say :
// Can not add historical input data(csv format) to Datastream when time zone is missing

Comment (aniket-amrutkar) [reply test/FalkonryTests.cs]: Please correct all the comments accordingly

File: test/FalkonryTests.cs [hunk 6]
```diff
@@ -953,27 +1112,184 @@ public void AddDataNarrowFormatCsv()
...
+            {
+                Assert.AreEqual(exception.Message, null, false);
+            }
+
+        }
 
     }
 
-    // [TestClass()]
+     [TestClass()]
```
Comment (aniket-amrutkar) [inline test/FalkonryTests.cs]: comment it

File: test/FalkonryTests.cs [hunk 7]
```diff
@@ -953,27 +1112,184 @@ public void AddDataNarrowFormatCsv()
...
+            catch (System.Exception exception)
+            {
+                Assert.AreEqual(exception.Message, "Missing entity Identifier.", true);
+            }
+
+        }
+
+        [TestMethod()]
+        // Add historical input data(csv format) to Datastream(Used for model revision)
+        public void AddDataCsvDifferentTimeZone()
```
Comment (aniket-amrutkar) [inline test/FalkonryTests.cs]: AddDataCsvDifferentTimeFormat

Comment (aniket-amrutkar) [reply test/FalkonryTests.cs]: Add two more test cases where , 
timeIDentifier and entityIdentifier are different than the default datastream values.

File: test/FalkonryTests.cs (hunk 3 above)
Comment (dilpeshjain) [reply test/FalkonryTests.cs]: This one is multi thing datastream.

Comment (aniket-amrutkar) [reply test/FalkonryTests.cs]: Ok

File: test/FalkonryTests.cs [hunk 8]
```diff
@@ -1081,7 +1081,7 @@ public void AddDataCsvMissingTimeIdentifier()
         }
 
         [TestMethod()]
-        // Add historical input data(csv format) to Datastream with time identifier missing
+        // Cannot add historical input data(csv format) to Datastream with time identifier missing
```
Comment (aniket-amrutkar) [inline test/FalkonryTests.cs]: change the comment.
add appropriate missing field in the comment
like timezone, time format etc

**Comment by aniket-amrutkar** [comment]
@dilpeshjain 
Also add dll to PR

File: README.md [hunk 9]
```diff
@@ -449,37 +449,74 @@ Falkonry C# Client to access [Falkonry Condition Prediction](falkonry.com) APIs
...
+    var options = new SortedDictionary<string, string>();
+    options.Add("streaming", "false");
+    options.Add("hasMoreData", "false");
     options.Add("timeIdentifier", "time");
-    options.Add("timeFormat", "YYYY-MM-DD HH:mm:ss");
-    options.Add("fileFormat", "json");
-	options.Add("streaming", "false");
-	options.Add("hasMoreData", "false");
-    InputStatus inputstatus = falkonry.addInput('datastream-id', data, options);
+    options.Add("timeZone", time.Zone);
```
Comment (phagunbaya) [inline README.md]: What is `time.zone` and `time.format`

Comment (dilpeshjain) [reply README.md]: In some test cases, we have defined time object, but here it isn't. Removed it.

**Pull request merged**